content/cpi-api-v2.md

```diff
@@ -0,0 +1,78 @@
+# Cloud Provider Interface (Version 2)
+
+For an overview of the sequence of CPI calls, the following resources are helpful:
+
+- [BOSH components](bosh-components.md) and its example component interaction diagram
+- [CLI v2 architecture doc](https://github.com/cloudfoundry/bosh-cli/blob/master/docs/architecture.md#deploy-command-flow) and [`bosh create-env` flow](https://github.com/cloudfoundry/bosh-init/blob/master/docs/init-cli-flow.png) where calls to the CPI are marked as `cloud`.
+
+Examples of API request and response:
+
+- [Building a CPI: RPC - Request](https://bosh.io/docs/build-cpi.html#request)
+- [Building a CPI: RPC - Response](https://bosh.io/docs/build-cpi.html#response)
+
+
+Library:
+
+- Ruby: `bosh-cpi-ruby` gem [v2.5.0](https://github.com/cloudfoundry/bosh-cpi-ruby/releases/tag/v2.5.0)
+- GoLang: `bosh-cpi-go` [library](https://github.com/cppforlife/bosh-cpi-go)
+
+#### Migration from V1 contracts
+
+Detailed instructions on how to migrate from V1 to V2 contracts can be found [here](v2-migration-guide.md).
+
+---
+
+## Glossary {: #glossary }
+
+- **cloud ID** is an ID (string) that the Director uses to reference any created infrastructure resource; typically CPI methods return cloud IDs and later receive them. For example AWS CPI's `create_vm` method would return `i-f789df` and `attach_disk` would take it.
+
+- **cloud_properties** is a hash that can be specified for several objects (resource pool, disk pool, stemcell, network) to provide infrastructure specific settings to the CPI for that object. Only CPIs know the meaning of its contents. For example resource pool's `cloud_properties` for AWS can specify `instance_type`:
+
+```yaml
+resource_pools:
+- name: large_machines
+  cloud_properties: {instance_type: r3.8xlarge}
+```
+
+## Methods
+
+- All V1 contracts must still be supported. See [CPI API V1](cpi-api-v1.md).
+- To differentiate V2 calls the caller needs to pass in `"api_version": 2` in the header of the request.
+
+### Reference Table (Based on each component version)
+
+| Director | CPI | Stemcell  | Should update registry   | Add agent setting to IaaS `user-metadata`   |
```

Comment:
Does `user-metadata` mean https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/master/stemcell_builder/stages/bosh_azure_agent_settings/apply.sh#L31 ?

According to https://github.com/cloudfoundry/bosh-notes/blob/master/proposals/cpi-api-v2.md, the disk settings are directly returned by CPI method. This solution is not used in the final implement. Correct? According to this new doc, the agent setting (including disk setting) is added to `user-metadata`, and bosh-agent will read from that.

Reply:
@bingosummer the CPI does return the disk attach point in `attach_disk` for V2. This is then used by the director to send the appropriate call to the agent (`mount_disk`) which includes this returned disk location. The director <> agent communication is replacing the registry here. 

For user-metadata, this table represents what needs to be in the agent settings for bootstrapping. As of V2, any settings you would have previously placed in the registry should be merged with the existing settings already entering the user-metadata. For actions such as attaching, mounting, detaching disks, the director communicates with the agent directly instead of the registry as a middle-man. 

For example, with the registry the user-metadata may look like:
```json
{
  "dns": ...
  "networks": ...
   "registry":
   {"endpoint": ...}
}
```
which upon bootstrapping the agent consults the registry for the remainder of the settings.

Without the registry agent settings would resemble something like:
```json
{
  "dns": ...
  "networks": ...
   "agent_id": ...
   "vm": {"name": ..."},
   "disks": {
      "ephemeral":  "/dev/sda1",
      "persistent": {}
}
```
it's important to note that persistent disks are never in the agent settings upon bootstrapping. The director and agent communicate further to attach/mount persistent disks. The persistent disk settings are written to another file in the agent, and the agent consults this file along with the settings deposited in the `user-metadata`. We'll improve the docs around this process.